### PR TITLE
Possibility to configure of animation duration

### PR DESCRIPTION
**Pull request by jershell**
## Description
add the possibility for the setting of the animation duration

## Related PRs/Issues
https://github.com/cortinico/slidetoact/issues/82

## Todos
- [X] Tests
- [X] Documentation

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -147,6 +148,18 @@ Use the ``slider_height`` attribute to control the **desired height** of the wid
 
 Use the ``slider_locked`` attribute to **lock the slider** (this is a boolean attribute). When a slider is locked, will always bump the button to the beginning (default is false).
 
+#### ``animation_duration``
+
+Use the ``animation_duration`` attribute to **set the duration**
```
Comment (cortinico) [inline README.md]:
```suggestion
Use the ``animation_duration`` attribute to **set the duration** of the complete and reset animation (in milliseconds).
```

File: README.md [hunk 2]
```diff
@@ -147,6 +148,18 @@ Use the ``slider_height`` attribute to control the **desired height** of the wid
...
+
```
Comment (cortinico) [inline README.md]: extra line please remove

File: example/src/main/res/layout/content_animation_duration.xml [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+        android:id="@+id/slide_container"
+        android:layout_width="match_parent"
+        android:layout_height="wrap_content"
+        android:clipToPadding="false"
+        android:orientation="vertical">
+
+        <com.ncorti.slidetoact.SlideToActView
+            style="@style/SlideToActView.Example"
+            app:animation_duration="2000"
+            app:text="duration 2000" />
```
Comment (cortinico) [inline example/src/main/res/layout/content_animation_duration.xml]: Please change the text to: `Animation Duration 2000`
to be more consistent with the other screens

File: slidetoact/src/main/java/com/ncorti/slidetoact/SlideToActView.kt [hunk 4]
```diff
@@ -103,6 +103,13 @@ class SlideToActView @JvmOverloads constructor (
             field = value
             mInnerPaint.color = value
             invalidate()
+        }    /** Inner color used by the slider (secondary, icon and border) */
```
Comment (cortinico) [inline slidetoact/src/main/java/com/ncorti/slidetoact/SlideToActView.kt]: Copy and paste?

File: slidetoact/src/main/java/com/ncorti/slidetoact/SlideToActView.kt [hunk 5]
```diff
@@ -103,6 +103,13 @@ class SlideToActView @JvmOverloads constructor (
...
+
+    var animDuration: Long = 0
+        set(value) {
+            field = value
+            mAnimDuration = value
+            invalidate()
```
Comment (cortinico) [inline slidetoact/src/main/java/com/ncorti/slidetoact/SlideToActView.kt]: You don't need to call `invalidate()` here

File: slidetoact/src/main/java/com/ncorti/slidetoact/SlideToActView.kt [hunk 6]
```diff
@@ -103,6 +103,13 @@ class SlideToActView @JvmOverloads constructor (
             field = value
             mInnerPaint.color = value
             invalidate()
+        }    /** Inner color used by the slider (secondary, icon and border) */
+
```
Comment (cortinico) [inline slidetoact/src/main/java/com/ncorti/slidetoact/SlideToActView.kt]: Please add a KDoc comment on top of this property

File: slidetoact/src/main/java/com/ncorti/slidetoact/SlideToActView.kt [hunk 7]
```diff
@@ -103,6 +103,13 @@ class SlideToActView @JvmOverloads constructor (
...
+    var animDuration: Long = 0
```
Comment (cortinico) [inline slidetoact/src/main/java/com/ncorti/slidetoact/SlideToActView.kt]: = 300

**Comment by jershell** [comment]
I resolve all conversations. About a "tickAnimator". Sorry, it not clear for me

**Comment by cortinico** [comment]
My apologize. The `tickAnimator` is the animator responsible of showing the Tick animation once the slider is completed:

https://github.com/cortinico/slidetoact/blob/b7e2914a25eb4e7786eb85428d9f67f71228112b/slidetoact/src/main/java/com/ncorti/slidetoact/SlideToActView.kt#L550

Ideally this animation should be affected by the same property used to configure the animation duration.

**Comment by jershell** [comment]
Please look, I made it

**Comment by jershell** [comment]
I think this (https://github.com/cortinico/slidetoact/issues/15) issue maybe will be resolved after the PR. I did a test on "genymotion enu". What do you say?

**Comment by jershell** [comment]
> mAnimationDuration

mAnimationDuration - private member
animDuration - just setter

**Comment by jershell** [comment]
Something trouble?

**Comment by abhijeet-nkt** [comment]
animDuration is not just a setter, it can be used to get the value of it's field (note field = value). Having another copy like mAnimationDuration is just redundant. You're following the Java philosophy where animDuration would have been a setter function), but that's not how things should be in Kotlin. animDuration serves as a field and property getter/setter as well.

**Comment by jershell** [comment]
> @jershell animDuration is not just a setter, it can be used to get the value of it's field (note field = value). Having another copy like mAnimationDuration is just redundant. You're following the Java philosophy (where animDuration would have been a setter function), but that's not how things should be in Kotlin. animDuration serves as a field and property getter/setter as well.

Ok, do you can fix this and apply PR?

**Comment by abhijeet-nkt** [comment]
I've never made a pull request before, so I don't know much. And the correction is in your pending PR, so don't know how I'd do that. Sorry for sounding dumb, can you help me out ?

**Comment by jershell** [comment]
> I've never made a pull request before, so I don't know much. And the correction is in your pending PR, so don't know how I'd do that. Sorry for sounding dumb, can you help me out ?

Okay, pls, help me too. 
I have the property for set value programmatically
 
```kotlin
    /** Duration of the complete and reset animation (in milliseconds). */
    var animDuration: Long = 300
        set(value) {
            field = value
            mAnimDuration = value
        }
```
and I should convert it to a simple getter If I understood right. So?

```kotlin
    var animDuration: Long = 0
        get() = mAnimDuration
        set(value) {
            field = value
            mAnimDuration = value
        }
```

or change to set only?

```kotlin
    fun setAnimDuration(value: Long) {
        mAnimDuration = value
    }
```

but then how to take the current value?

**Comment by abhijeet-nkt** [comment]
@jershell
Firstly, the mAnimDuration is not required at all,
So all you need is,

```
var animDuration: Long = 300
    set(value) {
        field = value
        invalidate()
    }
}
```

This should be enough, remove the line which declares mAnimDuration, it's not required

So whenever we need to access this property, you could just say
```
animSet.duration = animDuration
```

Changing it's value would be just:
```
animDuration = /* Some value */
```

This internally calls the setter, so you don't need to handle variable and setter separately

**Comment by jershell** [comment]
> @jershell
> Firstly, the mAnimDuration is not required at all,
> So all you need is,
> 
> ```
> var animDuration: Long = 300
>     set(value) {
>         field = value
>         invalidate()
>     }
> }
> ```
> 
> This should be enough, remove the line which declares mAnimDuration, it's not required
> 
> So whenever we need to access this property, you could just say
> 
> ```
> animSet.duration = animDuration
> ```
> 
> Changing it's value would be just:
> 
> ```
> animDuration = /* Some value */
> ```
> 
> This internally calls the setter, so you don't need to handle variable and setter separately

So, I did it

**Comment by jershell** [comment]
okay. What now?

**Comment by abhijeet-nkt** [comment]
@jarshell wait for him to merge into master

**Comment by cortinico** [comment]
Hi @jershell 
First, apologize for the major delay but I was on vacation and I turned off GH notifications.

Thanks for taking care of implementing what I asked:

> Ideally this animation should be affected by the same property used to configure the animation duration.

Unfortunately I realized that your implementation (specifically 39d078c) is removing the `AnimatedVectorDrawable` used to draw the tick from SlideToAct. The approach you used is really smart:

```kotlin
    /** Path for the tick */
    private val mRoutePath = Path().apply {
        moveTo(72.1F, 253.5f)
        rLineTo(103F, 103.7f)
        rLineTo(232.9f, -234.3f)
    }
```

On the other hand, this is not so flexible. I'd prefer having an `AnimatedVectorDrawable` inside the code. That will allow developers to customize the icon animation as they want (by just overriding resource files). 
I guess in this case is ok to sacrifice the animation duration of the tick for flexibility.

Long story short: I'd prefer to merge this PR at the commit fd6258b. Can I ask you to revert your changes and rebase on top of master (I can do it as well if you're busy).

Thanks again for your time :)

**Comment by jershell** [comment]
> Hi @jershell
> ...
> Long story short: I'd prefer to merge this PR at the commit [fd6258b](https://github.com/cortinico/slidetoact/commit/fd6258b211af93ab922649f7a4acb85b64f32cd7)
> 
> Thanks again for your time :)

Ok, I reset at commit fd6258b and merge with a fresh master
